### PR TITLE
[JENKINS-25972] Add enable/disable project action in breadcrumb menu

### PR DESCRIPTION
**Pull request by Austry**
See [JENKINS-25972](https://issues.jenkins-ci.org/browse/JENKINS-25972).

New link available in drop down menus [demo1](https://yadi.sk/i/Q6Gctfvz3P2iMp) [demo2](https://yadi.sk/i/VSTiKFJV3P2iMn)
Also available in project configuration [demo](https://yadi.sk/i/tTsFhl9G3P2iMt)

Not sure about icons choice

Don't know if this feature requires UI tests or not
### Proposed changelog entries

* Added "Disable Project"/"Enable Project" controls to drop down menus

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

### Desired reviewers

@jenkinsci/hacktoberfest

### CODE REVIEW

File: core/src/main/java/hudson/model/FreeStyleProject.java [hunk 1]
```diff
@@ -56,6 +56,7 @@ public FreeStyleProject(ItemGroup parent, String name) {
         return FreeStyleBuild.class;
     }
 
+    @Override
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/FreeStyleProject.java]: Please don't include general cleanup in otherwise unrelated files in your PR.

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java [hunk 2]
```diff
@@ -128,6 +128,19 @@ public ContextMenu add(String url, String icon, String text, boolean post, boole
             return this;
         }
 
+        /** @since 2.87 */
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: It may take us a while to merge something, so the safer approach is to `@since TODO` and let this be inserted on merge. Better TODO than an altogether wrong version because someone missed it.

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java [hunk 3]
```diff
@@ -128,6 +128,19 @@ public ContextMenu add(String url, String icon, String text, boolean post, boole
...
+        public ContextMenu add(String url, String icon, String text, boolean post, boolean requiresConfirmation,
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: The other `add` methods probably should start calling this one to limit the code duplication.

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java [hunk 4]
```diff
@@ -259,6 +272,12 @@ public void run(JellyContext context, XMLOutput output) throws JellyTagException
          */
         @Exported public boolean requiresConfirmation;
 
+         /**
+         * True to require menu reload after a click.
+         * @since 2.87
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: See above.

File: core/src/main/resources/lib/hudson/project/configurable.jelly [hunk 5]
```diff
@@ -45,4 +45,22 @@ THE SOFTWARE.
...
+                <l:task href="${url}/enable" post="true" icon="icon-next icon-md" title="${%Enable build}" onclick="return postAndReload(this)" permission="${it.CONFIGURE}" requestsMenuInvalidate="true"/>
+            </j:when>
+            <j:otherwise>
+                <l:task href="${url}/disable" post="true" icon="icon-folder-delete icon-md" title="${%Disable build}" onclick="return postAndReload(this)" permission="${it.CONFIGURE}" requestsMenuInvalidate="true"/>
+            </j:otherwise>
+        </j:choose>
+        <script>
+            function postAndReload(a) {
+                new Ajax.Request(a.href);
+                location.reload();
```
Comment (daniel-beck) [inline core/src/main/resources/lib/hudson/project/configurable.jelly]: This seems weird considering that the context menu of an item can be accessed from other views, e.g. a list view.

Does this even work when accessing this link from a context menu on another view that doesn't have this snippet?

File: core/src/main/resources/lib/layout/task.jelly [hunk 6]
```diff
@@ -72,6 +72,11 @@ THE SOFTWARE.
       Message to use for confirmation, if requested; defaults to title.
       (since 1.512)
     </st:attribute>
+    <st:attribute name="requestsMenuInvalidate" type="boolean">
+      Request menu reload after a click.
+      Used for reloading breadcrumb menu then project state changed.
+      (since 2.87)
```
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/task.jelly]: See above.

**Comment by daniel-beck** [comment]
Apparently there's a duplicate issue here that nobody noticed, as @olivergondza proposed to implement this (as a different issue) in #1703. Perhaps you can gain something from the comment thread over there (other than that some are not convinced this is a good idea).

Personally I'd prefer an implementation based on `TransientAction` and `TransientActionFactory`, as these are much nicer than hardcoded entries.

File: core/src/main/resources/lib/hudson/project/configurable.jelly (hunk 5 above)
Comment (Austry) [reply core/src/main/resources/lib/hudson/project/configurable.jelly]: It's not included for context menus
It's there for reloading the page them you disable project on project configuration page. With this you will see updated UI after click.

For drop down menus this page(configurable.jelly) transforms into JSON object with <j:tasks> items and theirs properties and click on it proceeded in breadcrumbs.js.

**Comment by Austry** [comment]
PR updated
@daniel-beck comments fixed

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java [hunk 7]
```diff
@@ -259,6 +259,12 @@ public void run(JellyContext context, XMLOutput output) throws JellyTagException
          */
         @Exported public boolean requiresConfirmation;
 
+         /**
+         * True to require menu reload after a click.
+         * @since TODO
+         */
+        @Exported public boolean requestsMenuInvalidate;
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: Not sure if it's required in REST API, but OK

Comment (oleg-nenashev) [reply core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: OTOH I would advice against adding new public non-final fields. Better to have a private field and a getter

File: core/src/main/resources/lib/layout/breadcrumbs.js [hunk 8]
```diff
@@ -49,11 +49,14 @@ var breadcrumbs = (function() {
         return a+'/'+b+qs;
     }
 
-    function postRequest(action, event, url) {
-        new Ajax.Request(url);
+    function postRequest(action, event, cfg) {
```
Comment (oleg-nenashev) [inline core/src/main/resources/lib/layout/breadcrumbs.js]: any chance it's used somewhere else? I am not sure whether it may break compatibility

File: core/src/main/resources/lib/layout/breadcrumbs.js [hunk 9]
```diff
@@ -202,16 +205,16 @@ var breadcrumbs = (function() {
...
-                        e.text = makeMenuHtml(e.icon, e.displayName);
-                        if (e.subMenu!=null)
-                            e.subMenu = {id:"submenu"+(iota++), itemdata:e.subMenu.items.each(fillMenuItem)};
-                        if (e.requiresConfirmation) {
-                            e.onclick = {fn: requireConfirmation, obj: {url: e.url, displayName: e.displayName, post: e.post}};
-                            delete e.url;
-                        } else if (e.post) {
-                            e.onclick = {fn: postRequest, obj: e.url};
-                            delete e.url;
+                    function fillMenuItem(item) {
```
Comment (oleg-nenashev) [inline core/src/main/resources/lib/layout/breadcrumbs.js]: It would be great to do refactorings in separate pull requests. It complicates reviews of changes

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java (hunk 7 above)
Comment (Austry) [reply core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: but what about consistency? 
do you think it will be better to refactor all fields of  ```MenuItem``` to private final?
in different PR

File: core/src/main/resources/lib/layout/breadcrumbs.js (hunk 8 above)
Comment (Austry) [reply core/src/main/resources/lib/layout/breadcrumbs.js]: not found any over usage

File: core/src/main/resources/lib/layout/breadcrumbs.js (hunk 9 above)
Comment (Austry) [reply core/src/main/resources/lib/layout/breadcrumbs.js]: yeah, sorry

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java (hunk 7 above)
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: Refactoring is not trivial, these fields are the part of the public API. I suggest to just avoid growing the technical debt

File: core/src/main/resources/lib/layout/breadcrumbs.js (hunk 8 above)
Comment (oleg-nenashev) [reply core/src/main/resources/lib/layout/breadcrumbs.js]: @tfennelly @kzantow @recena Do you see any risks?

**Review by oleg-nenashev — APPROVED** [review]
LGTM 👍

File: core/src/main/resources/lib/hudson/project/configurable.jelly [hunk 10]
```diff
@@ -45,4 +45,22 @@ THE SOFTWARE.
             <l:task href="${url}/configure" icon="icon-gear2 icon-md" title="${%View Configuration}"/>
         </j:when>
     </j:choose>
+    <j:if test="${it.supportsMakeDisabled()}">
+        <j:choose>
+            <j:when test="${it.disabled}">
+                <l:task href="${url}/enable" post="true" icon="icon-next icon-md" title="${%Enable build}" onclick="return postAndReload(this)" permission="${it.CONFIGURE}" requestsMenuInvalidate="true"/>
+            </j:when>
+            <j:otherwise>
+                <l:task href="${url}/disable" post="true" icon="icon-folder-delete icon-md" title="${%Disable build}" onclick="return postAndReload(this)" permission="${it.CONFIGURE}" requestsMenuInvalidate="true"/>
```
Comment (jglick) [inline core/src/main/resources/lib/hudson/project/configurable.jelly]: These icons seems rather unintuitive to me. The only existing GUI we have for disabled projects is a grey ball, and these match neither. Definitely a folder icon is weird for a non-folder.

Maybe we could just embed a checkbox in the menu item? Or something akin to the outline-vs.-colored icons used by the `favorite` plugin?

**Review by jglick — COMMENTED** [review]
The GUI seems like it needs some design work here.

Not really sure it is worth putting significant effort into Classic UI when there is already a working enable/disable gesture, and streamlined UIs could be done in Blue Ocean.

File: core/src/main/resources/lib/hudson/project/configurable.jelly (hunk 10 above)
Comment (daniel-beck) [reply core/src/main/resources/lib/hudson/project/configurable.jelly]: Pause/play might work, and are in Tango, but could lead to confusion with "Build Now".

**Comment by oleg-nenashev** [comment]
@Austry @jglick @daniel-beck @jenkinsci/code-reviewers Regarding the design feedback above, how do we proceed?

**Comment by recena** [comment]
I agree with @jglick, the icons are not proper.

**Comment by Austry** [comment]
sorry for long answer
I think green/grey ball more suitable here
pause/play icons lead to confusion with "Build Now" in my opinion

**Comment by daniel-beck** [comment]
This is probably supposed to be `icon-blue`.

Not sure that makes sense though, as those are mostly build statuses per http://ci.jenkins.io/legend

**Review by daniel-beck — DISMISSED** [review]
I hate to write this, but reusing the build status icons for enabling/disabling for jobs doesn't seem right.

There needs to be a better alternative.

@i386?

**Comment by oleg-nenashev** [comment]
So, how do we proceed here?

**Comment by daniel-beck** [comment]
Well, the PR won't work as is. Using the same icon for two wildly different behaviors doesn't seem like a good idea.

When I introduced the global tools configuration, I moved icons around. If necessary, something similar could be done here. But I don't think this is needed. Could be just using `player_pause` and `player_play` from Tango, or `play` and `pause` from the RRZE icon set.

**Comment by Austry** [comment]
I change play/pause icons to icons from [RRZE](https://rrze-pp.github.io/rrze-icon-set/tango/gallery.html)
here how it looks [play](https://yadi.sk/i/1LYmfPjs3S5k45) [pause](https://yadi.sk/i/L3sQNyFR3S5jzp)
may be green/yellow icons would be better? [1](https://raw.githubusercontent.com/RRZE-PP/rrze-icon-set/master/tango/32x32/actions/pause-yellow.png) [2](https://raw.githubusercontent.com/RRZE-PP/rrze-icon-set/master/tango/32x32/actions/pause-yellow.png)

**Comment by oleg-nenashev** [comment]
There is a related discussion in #3320. Once there is a decision there, the same icon can be applied in the PR so that we finally merge it

**Comment by timja** [comment]
Looks very stale, closing to clean up the project. It was proposed for close over a month ago

**Comment by oleg-nenashev** [comment]
Could we keep it open for a while? Closing tag was added without justification, and the PR itself is still in a healthy state

**Comment by rsandell** [comment]
Maybe can be merged for this year's hacktoberfest? :)

**Comment by Austry** [comment]
I'm ready to continue work on this. But do not know that to do.
Is the last problem is icons?

File: core/src/main/resources/lib/layout/task.jelly [hunk 11]
```diff
@@ -72,6 +72,11 @@ THE SOFTWARE.
       Message to use for confirmation, if requested; defaults to title.
       (since 1.512)
     </st:attribute>
+    <st:attribute name="requestsMenuInvalidate" type="boolean">
+      Request menu reload after a click.
+      Used for reloading breadcrumb menu then project state changed.
```
Comment (varyvol) [inline core/src/main/resources/lib/layout/task.jelly]:
```suggestion
      Used to reload breadcrumb menu when the project state changes.
```

**Comment by varyvol** [comment]
@oleg-nenashev this was reopened and was waiting for another PR that was eventually closed. What's your opinion about this now? Can we propose some actions for @Austry to move this forward? Thanks!

**Comment by MRamonLeon** [comment]
@Austry there is no agreement on the PR @oleg-nenashev commented (https://github.com/jenkinsci/jenkins/pull/3320), it's closed. I will be happy to approve it after you address @varyvol suggestions (some typos)

**Comment by varyvol** [comment]
@Austry would you be interested in moving this forward? You could apply the suggestions if you think they are fine and maybe try to gather feedback (from Oleg and from others) in the mailing list.

**Comment by Austry** [comment]
I am not interested anymore. You can pick it up. Sorry.

**Comment by fcojfernandez** [comment]
Closing per author's reply
